### PR TITLE
Wrap eth v2

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Continues the work on and supersedes https://github.com/gnosis/gp-swap-ui/pull/72

Implemented wrapping ETH when selling.

When `from` is set to `ETH`, 2 wallet pop ups should be displayed in sequence to the user:

1. Wrap necessary amount to complete the order of `ETH` into `WETH`. Onchain
2. Place order. Offchain

screenshot_2021-01-08_13-33-56

Works both with fixed `from` ("sell" order) and fixed `to` ("buy" order) amounts.

# To test

1. Fire the app
2. Select ETH on `from`, and anything else except WETH on `to`
3. Click `Swap` and confirm.

- [ ] The first pop up should be the one to wrap ETH
- [ ] The second pop up should be to sign the offline order

# Caveats

- :warning:  Does NOT unwrap WETH when receiving ETH (`ETH` set on `to` field)
- If you cancel the first request, you'll still be prompted for, and can sign the offline order

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
   * **🎩 [Dapp](https://pr95--gpswapui.review.gnosisdev.com)**: Testing web app

File: src/hooks/useWrapEther.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { CurrencyAmount } from '@uniswap/sdk'
+
+import { useTransactionAdder } from '@src/state/transactions/hooks'
+
+import { Contract } from 'ethers'
+
+export function useWrapEther() {
+  const addTransaction = useTransactionAdder()
+
+  return async (amount: CurrencyAmount, weth: Contract): Promise<string> => {
```
Comment (W3stside) [inline src/hooks/useWrapEther.ts:10]: can you return a `useCallback` here to memoise?

File: src/custom/hooks/useSwapCallback.ts [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
 import { BigNumber } from 'ethers'
-import { Percent, Trade, TradeType } from '@uniswap/sdk'
 
+import { ETHER, Percent, Trade, TradeType } from '@uniswap/sdk'
 import { useActiveWeb3React } from '@src/hooks'
 import useENS from '@src/hooks/useENS'
 import { useMemo } from 'react'
 import useTransactionDeadline from '@src/hooks/useTransactionDeadline'
+
+import { computeSlippageAdjustedAmounts } from '@src/utils/prices'
```
Comment (W3stside) [inline src/custom/hooks/useSwapCallback.ts:12]: i don't think we need every import here with `@src` prepended, i think

**Review by W3stside — APPROVED** [review]

File: src/custom/hooks/useSwapCallback.ts (hunk 2 above)
Comment (alfetopito) [reply src/custom/hooks/useSwapCallback.ts:12]: Ok.

Why do we have `@src` at all, BTW? Does it have to do with the custom path thingy?

File: src/custom/hooks/useSwapCallback.ts [hunk 3]
```diff
@@ -31,9 +34,11 @@ export function useSwapCallback(
   const validTo = useTransactionDeadline()?.toNumber() || MAX_VALID_TO_EPOCH
   const addPendingOrder = useAddPendingOrder()
   const { INPUT: inputAmount, OUTPUT: outputAmount } = computeSlippageAdjustedAmounts(trade, allowedSlippage)
+  const weth = useWETHContract()
+  const wrapEther = useWrapEther()
 
   return useMemo(() => {
-    if (!trade || !library || !account || !chainId || !inputAmount || !outputAmount) {
+    if (!trade || !library || !account || !chainId || !inputAmount || !outputAmount || !weth) {
```
Comment (anxolin) [inline src/custom/hooks/useSwapCallback.ts:41]: I guess that here, u would need to check that wrapEther is not null (I understand returns null if WETH is not available...) Related to my other comment.

**Review by anxolin — APPROVED** [review]
Good PR, as we discussed in the sync, this functionality can be improved UX wise but I think this is a pretty good start! 

Next steps regarding selling ETH, I would just try to discuss options on how to inform better the user, or alternative ways. It would be nice to review what are the issues with this option, and see alternatives (even if we end up to leave it like this for a while)

File: src/hooks/useWrapEther.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/hooks/useWrapEther.ts:10]: I was going to do that, but then we have to handle the error when `weth` is not available, and this hook initially doesn't return any errors...
But yeah, makes sense, will change it

File: src/custom/hooks/useSwapCallback.ts (hunk 2 above)
Comment (alfetopito) [reply src/custom/hooks/useSwapCallback.ts:12]: So, I removed all `@src/` import prefixes and it didn't quite workout.
Got some weird errors regarding import order on `useContracts`...

Anyway, imports look quite messy on this file.
I'll try to sort them out.

**Comment by W3stside** [comment]
`@src` refers to the original `src` files. if there are custom files you
use the normal route without `@src` to go for custom path

**Pull request merged**